File: androidtv/app/src/main/java/tv/enktel/app/data/repo/SetupLink.kt
package tv.enktel.app.data.repo

import java.net.URLDecoder

/**
 * Works out what a subscriber just pasted.
 *
 * ## Why this exists
 *
 * The setup form asked, in order: Xtream Codes or M3U Playlist; a playlist
 * name; a server URL "(http://host:port)"; a username; a password; and an
 * optional film-library M3U. Six decisions, the first of which is a protocol
 * question, before anyone can watch anything. The feedback was that the app is
 * hard to set up, and this form is most of why.
 *
 * None of those questions are ones the subscriber has an answer to. What they
 * *have* is whatever their provider sent them — for our own subscribers, the
 * welcome email, which carries a server address, a username, a password and an
 * M3U link. Every one of those already says which kind of line it is and, in
 * the M3U case, carries the credentials in its query string.
 *
 * So the form can ask for one thing: paste it. This turns the paste into a
 * [Setup].
 *
 * ## What it accepts
 *
 * - `http://host:8080/get.php?username=u&password=p&type=m3u_plus` — the M3U
 *   link from a welcome email. Yields an **Xtream** line, not an M3U one: the
 *   credentials are right there, and an Xtream profile gets categories, VOD,
 *   series and catch-up that a flat playlist cannot.
 * - `http://host:8080/player_api.php?username=u&password=p` — same.
 * - `http://host:8080` plus a username and password typed separately.
 * - Any other playlist URL (`.m3u`, `.m3u8`, an IPTV-Editor or Xtream-proxy
 *   link) — an M3U line.
 * - The welcome email's credentials block pasted whole, newlines and labels
 *   and all:
 *
 *       Username: enktel_demo
 *       Password: 8fj3kd92
 *       Server address: http://host:8080
 *
 * ## What it deliberately does not do
 *
 * It does not reach the network. Deciding what something *is* and finding out
 * whether it *works* are different jobs, and only the second one can fail
 * slowly or leave a half-made profile behind. [PlaylistRepository.addXtream]
 * still validates, still probes both schemes, and still reports the panel's
 * own refusal.
 */
object SetupLink {

    /** What a paste turned out to be. */
    sealed interface Setup {
        /** A panel line: server plus credentials. */
        data class Xtream(
            val server: String,
            val username: String,
            val password: String,
        ) : Setup

        /** A playlist URL, with no credentials to pull out of it. */
        data class M3u(val url: String) : Setup

        /**
         * A server was recognised but no credentials came with it.
         *
         * Not a failure: it is the normal state of the form after someone has
         * pasted a bare host, and the caller's job is to ask for the two
         * fields rather than to complain.
         */
        data class NeedsCredentials(val server: String) : Setup

        /** Nothing usable. [reason] is written to be shown to a subscriber. */
        data class Unrecognised(val reason: String) : Setup
    }

    /** Endpoints whose path carries credentials we can lift. */
    private val CREDENTIALLED = listOf("get.php", "player_api.php", "panel_api.php", "xmltv.php")

    /**
     * Labels the welcome email and most providers use, lowercased.
     *
     * Matched on a line's prefix rather than searched for anywhere in it: a
     * password can contain the word "user", and a line-leading label is the
     * only position that reliably means "this is the field name".
     */
    private val USER_LABELS = listOf("username", "user name", "user", "login")
    private val PASS_LABELS = listOf("password", "pass", "pwd")
    private val SERVER_LABELS = listOf("server address", "server url", "server", "host", "portal", "url", "dns")

    /**
     * Parse a paste, optionally with a username and password the subscriber
     * typed into their own fields.
     *
     * [typedUser] and [typedPass] win over anything found in the text: if
     * someone has filled the fields in, that is a more recent statement of
     * intent than whatever the pasted link happens to contain.
     */
    fun parse(raw: String, typedUser: String = "", typedPass: String = ""): Setup {
        val text = raw.trim()
        if (text.isBlank() && typedUser.isBlank()) {
            return Setup.Unrecognised("Paste the link or details your provider sent you.")
        }

        val fromLabels = readLabelled(text)
        val url = firstUrl(text)

        val user = typedUser.trim().ifBlank { fromLabels.user }
        val pass = typedPass.trim().ifBlank { fromLabels.pass }

        // A credentialled endpoint is the best case: it names the host and
        // carries the login, so nothing has to be typed at all.
        if (url != null && isCredentialled(url)) {
            val q = queryParams(url)
            val u = user.ifBlank { q["username"].orEmpty() }
            val p = pass.ifBlank { q["password"].orEmpty() }
            val server = PlaylistRepository.normalizeServer(url)
            return if (u.isNotBlank() && p.isNotBlank()) {
                Setup.Xtream(server, u, p)
            } else {
                // The shape said Xtream but the credentials are not in it —
                // an edited link, or one whose parameters are named something
                // else. Ask for the two fields rather than guessing.
                Setup.NeedsCredentials(server)
            }
        }

        // A plain playlist link. Only treated as M3U when there is no username
        // to pair it with: a subscriber who pasted a playlist *and* typed a
        // login is describing a panel, and a panel line is the better profile.
        if (url != null && looksLikePlaylist(url) && user.isBlank()) {
            return Setup.M3u(url)
        }

        val server = url?.let { PlaylistRepository.normalizeServer(it) }
            ?: fromLabels.server.takeIf { it.isNotBlank() }?.let { PlaylistRepository.normalizeServer(it) }

        if (server != null) {
            return if (user.isNotBlank() && pass.isNotBlank()) {
                Setup.Xtream(server, user, pass)
            } else {
                Setup.NeedsCredentials(server)
            }
        }

        // No address anywhere. Said in terms of what to do next rather than
        // what failed — "could not parse input" helps nobody holding a phone.
        return Setup.Unrecognised(
            "That doesn't look like a playlist link. Paste the whole line your " +
                "provider sent, starting with http.",
        )
    }

    /** A sensible profile name, so nobody has to invent one. */
    fun suggestedName(setup: Setup): String = when (setup) {
        is Setup.Xtream -> hostOf(setup.server)
        is Setup.M3u -> hostOf(setup.url)
        is Setup.NeedsCredentials -> hostOf(setup.server)
        is Setup.Unrecognised -> "My Playlist"
    }

    private fun hostOf(url: String): String =
        runCatching {
            java.net.URI(if (url.contains("://")) url else "http://$url").host
                ?.removePrefix("www.")
                ?.takeIf { it.isNotBlank() }
        }.getOrNull() ?: "My Playlist"

    // ── the pieces ────────────────────────────────────────────────────

    private data class Labelled(val user: String, val pass: String, val server: String)

    /**
     * Pull `Username: x` / `Password: y` / `Server: z` out of pasted text.
     *
     * Split on the first `:` or `=` after the label, because "Server address:
     * http://host:8080" contains three colons and only the first one is the
     * separator.
     */
    private fun readLabelled(text: String): Labelled {
        var user = ""
        var pass = ""
        var server = ""
        for (line in text.lines()) {
            val trimmed = line.trim()
            if (trimmed.isEmpty()) continue
            val cut = trimmed.indexOfFirst { it == ':' || it == '=' }
            if (cut <= 0) continue
            val label = trimmed.take(cut).trim().lowercase().trimEnd('*', '-', '•', ' ')
            val value = trimmed.substring(cut + 1).trim()
            if (value.isBlank()) continue
            when {
                // Checked longest-first within each set so "user name" is not
                // eaten by "user".
                user.isBlank() && USER_LABELS.any { label == it } -> user = value
                pass.isBlank() && PASS_LABELS.any { label == it } -> pass = value
                server.isBlank() && SERVER_LABELS.any { label == it } -> server = value
            }
        }
        return Labelled(user, pass, server)
    }

    /** The first http(s) address in the text, or null. */
    private fun firstUrl(text: String): String? =
        Regex("""https?://\S+""", RegexOption.IGNORE_CASE)
            .find(text)
            ?.value
            // Trailing punctuation from prose — "…&output=ts." or a link at
            // the end of a sentence inside a quoted email.
            ?.trimEnd('.', ',', ';', ')', '>', '"', '\'')

    /**
     * Is this one of the panel endpoints, whether or not it carries a login?
     *
     * Deliberately not `&& contains("username=")`. A `/get.php` link with its
     * credentials stripped is still a panel address — it is simply missing the
     * two fields — and treating it as a flat playlist instead would build an
     * M3U profile that can never load, off a URL that plainly names a panel.
     * The caller returns [Setup.NeedsCredentials] for that case.
     */
    private fun isCredentialled(url: String): Boolean {
        val lower = url.lowercase()
        return CREDENTIALLED.any { lower.contains("/$it") }
    }

    /**
     * Does this look like a playlist rather than a panel address?
     *
     * A path or a query is the tell: `http://host:8080` is a panel, whereas
     * anything with something after the host is a link to a specific file or
     * endpoint. `.m3u`/`.m3u8` are named explicitly because they are the
     * common case and worth being certain about.
     */
    private fun looksLikePlaylist(url: String): Boolean {
        val lower = url.lowercase()
        if (lower.contains(".m3u")) return true
        val afterScheme = lower.substringAfter("://")
        val path = afterScheme.substringAfter('/', "")
        return path.isNotBlank()
    }

    /**
     * Query parameters, decoded.
     *
     * Hand-rolled rather than `Uri.parse`: this is in `data/`, it is covered
     * by a plain JVM test, and pulling in the Android URI parser would make
     * that test need Robolectric for two string splits.
     */
    private fun queryParams(url: String): Map<String, String> {
        val query = url.substringAfter('?', "")
        if (query.isBlank()) return emptyMap()
        return query.split('&').mapNotNull { pair ->
            val k = pair.substringBefore('=', "")
            val v = pair.substringAfter('=', "")
            if (k.isBlank()) null else k.lowercase() to decode(v)
        }.toMap()
    }

    private fun decode(s: String): String =
        runCatching { URLDecoder.decode(s, "UTF-8") }.getOrDefault(s)
}

File: pc/src/lib/setupLink.ts
/**
 * Works out what someone just pasted into the setup box.
 *
 * A port of `androidtv/.../data/repo/SetupLink.kt`, deliberately kept
 * behaviour-for-behaviour identical. The two apps are handed the same welcome
 * email by the same subscriber, and a desktop that rejects what the phone
 * accepted is worse than either being strict — it reads as one of them being
 * broken. `setupLink.spec.mjs` runs the Kotlin test suite's cases against this
 * file for exactly that reason.
 *
 * The desktop form used to ask for Xtream-vs-M3U, a playlist name, a server
 * URL "(http://host:port)", a username and a password. The phone stopped
 * asking any of that at 1.69.0; this is the other half of that change.
 */

/** What a paste turned out to be. */
export type Setup =
  | { kind: 'xtream'; server: string; username: string; password: string }
  | { kind: 'm3u'; url: string }
  /** A host was recognised, the login was not. Ask for two fields. */
  | { kind: 'needsCredentials'; server: string }
  /** Nothing usable. `reason` is written to be shown to a subscriber. */
  | { kind: 'unrecognised'; reason: string };

/** Endpoints whose path names a panel, whether or not they carry a login. */
const CREDENTIALLED = ['get.php', 'player_api.php', 'panel_api.php', 'xmltv.php'];

const USER_LABELS = ['username', 'user name', 'user', 'login'];
const PASS_LABELS = ['password', 'pass', 'pwd'];
const SERVER_LABELS = ['server address', 'server url', 'server', 'host', 'portal', 'url', 'dns'];

/**
 * A server address reduced to scheme, host and port.
 *
 * The same rules as `PlaylistRepository.normalizeServer` on Android: strip a
 * pasted endpoint path and query, and when no scheme was given, infer one —
 * a bare host on a non-standard port is plain HTTP in this ecosystem, a bare
 * hostname is HTTPS. The two must agree, because a subscriber who set the
 * phone up by pasting one line will paste the same line here.
 */
export function normalizeServer(raw: string): string {
  let s = raw.trim().replace(/\/+$/, '');
  if (!s) return '';
  if (!/^https?:\/\//i.test(s)) {
    const host = s.split('/')[0];
    const port = host.includes(':') ? Number(host.split(':').pop()) : NaN;
    s = Number.isFinite(port) && port !== 443 ? `http://${s}` : `https://${s}`;
  }
  s = s
    .split('/player_api.php')[0]
    .split('/get.php')[0]
    .split('/panel_api.php')[0]
    .split('/xmltv.php')[0];
  return s.split('?')[0].replace(/\/+$/, '');
}

/**
 * @param defaultServer the host to assume when the paste names none. A
 *   reseller very often hands out a username and a password and nothing else;
 *   blank disables the fallback, because inventing a host for someone whose
 *   provider we do not know sends them to a stranger's panel.
 */
export function parseSetup(
  raw: string,
  typedUser = '',
  typedPass = '',
  defaultServer = '',
): Setup {
  const text = raw.trim();
  if (!text && !typedUser.trim()) {
    return { kind: 'unrecognised', reason: 'Paste the link or details your provider sent you.' };
  }

  const labelled = readLabelled(text);
  const url = firstUrl(text);

  const user = typedUser.trim() || labelled.user;
  const pass = typedPass.trim() || labelled.pass;

  // Best case: the address and the login in one string, nothing to type.
  if (url && isCredentialled(url)) {
    const q = queryParams(url);
    const u = user || q['username'] || '';
    const p = pass || q['password'] || '';
    const server = normalizeServer(url);
    return u && p
      ? { kind: 'xtream', server, username: u, password: p }
      : { kind: 'needsCredentials', server };
  }

  // A plain playlist link, only when there is no login to pair it with: a
  // panel profile gets categories, VOD, series and catch-up that a flat
  // playlist cannot.
  if (url && looksLikePlaylist(url) && !user) {
    return { kind: 'm3u', url };
  }

  const server = url
    ? normalizeServer(url)
    : labelled.server
      ? normalizeServer(labelled.server)
      : '';

  if (server) {
    return user && pass
      ? { kind: 'xtream', server, username: user, password: pass }
      : { kind: 'needsCredentials', server };
  }

  if (user && pass && defaultServer) {
    return { kind: 'xtream', server: normalizeServer(defaultServer), username: user, password: pass };
  }

  return {
    kind: 'unrecognised',
    reason:
      "That doesn't look like a playlist link. Paste the whole line your " +
      'provider sent, starting with http.',
  };
}

/** A sensible profile name, so nobody has to invent one. */
export function suggestedName(s: Setup): string {
  switch (s.kind) {
    case 'xtream':
    case 'needsCredentials':
      return hostOf(s.server);
    case 'm3u':
      return hostOf(s.url);
    default:
      return 'My playlist';
  }
}

/**
 * Turn a failure into something worth reading.
 *
 * Same wording as the Android app's `friendly()`. A subscriber who reads one
 * message on their phone and a different one here concludes the two apps
 * disagree about their account.
 */
export function friendlyError(message: string): string {
  const m = message || '';
  const has = (s: string) => m.toLowerCase().includes(s.toLowerCase());
  if (has('rejected the credentials') || has('invalid username') || has('unauthorised') || has('unauthorized')) {
    return "That username or password wasn't accepted. Check them for a stray space, then try again.";
  }
  if (has('unknownhost') || has('unable to resolve host') || has('enotfound') || has('getaddrinfo')) {
    return "Couldn't reach that address. Check your internet, and check the address for a typo.";
  }
  if (has('timeout') || has('timed out')) {
    return "The provider didn't answer in time. It may be busy — try again in a minute.";
  }
  if (has('certificate') || has('ssl') || has('tls')) {
    return "That server's security certificate couldn't be checked. Try http:// instead of https://.";
  }
  return m || "Couldn't connect. Check the link and try again.";
}

// ── the pieces ────────────────────────────────────────────────────────

function hostOf(url: string): string {
  try {
    const u = new URL(url.includes('://') ? url : `http://${url}`);
    return u.hostname.replace(/^www\./, '') || 'My playlist';
  } catch {
    return 'My playlist';
  }
}

/**
 * Pull `Username: x` / `Password: y` / `Server: z` out of pasted text.
 *
 * Split on the first `:` or `=`, because "Server address: http://host:8080"
 * contains three colons and only the first is the separator. Matched on a
 * line's leading label rather than searched for anywhere in it — a password
 * can contain the word "user".
 */
function readLabelled(text: string): { user: string; pass: string; server: string } {
  let user = '';
  let pass = '';
  let server = '';
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    const cut = [...trimmed].findIndex((c) => c === ':' || c === '=');
    if (cut <= 0) continue;
    const label = trimmed.slice(0, cut).trim().toLowerCase().replace(/[*\-•\s]+$/, '');
    const value = trimmed.slice(cut + 1).trim();
    if (!value) continue;
    if (!user && USER_LABELS.includes(label)) user = value;
    else if (!pass && PASS_LABELS.includes(label)) pass = value;
    else if (!server && SERVER_LABELS.includes(label)) server = value;
  }
  return { user, pass, server };
}

/** The first http(s) address in the text, stripped of trailing prose. */
function firstUrl(text: string): string | null {
  const m = /https?:\/\/\S+/i.exec(text);
  if (!m) return null;
  return m[0].replace(/[.,;)>"']+$/, '');
}

function isCredentialled(url: string): boolean {
  const lower = url.toLowerCase();
  return CREDENTIALLED.some((e) => lower.includes(`/${e}`));
}

function looksLikePlaylist(url: string): boolean {
  const lower = url.toLowerCase();
  if (lower.includes('.m3u')) return true;
  const path = lower.split('://')[1]?.split('/').slice(1).join('/') ?? '';
  return path.length > 0;
}

function queryParams(url: string): Record<string, string> {
  const q = url.split('?')[1];
  if (!q) return {};
  const out: Record<string, string> = {};
  for (const pair of q.split('&')) {
    const k = pair.split('=')[0];
    if (!k) continue;
    const v = pair.slice(k.length + 1);
    try {
      out[k.toLowerCase()] = decodeURIComponent(v);
    } catch {
      out[k.toLowerCase()] = v;
    }
  }
  return out;
}
